File: Columna para números decimales en DataGridView/[C#]-Columna para números decimales en DataGridView/C#/GridExtension/NumericGridColumn.cs
using System;
using System.ComponentModel;
using System.Globalization;
using System.Windows.Forms;

namespace GridExtension
{
    public class NumericGridColumn : DataGridViewColumn
    {
        private NumberFormatInfo _numberFormat;
        private string _decimalSeparator;
        private int _decimalDigits = -1;

        public NumericGridColumn():base (new NumericGridCell()) { }

        public override DataGridViewCell CellTemplate
        {
            get
            {
                return base.CellTemplate;
            }
            set
            {
                if (value != null &&
                        !value.GetType().IsAssignableFrom(typeof(NumericGridCell)))
                    throw new InvalidCastException("Debe especificar una instancia de NumericGridCell");
                base.CellTemplate = value;
            }
        }

        [Category("Appearance")]
        public string DecimalSeparator {
            get {
                if (string.IsNullOrEmpty(_decimalSeparator))
                    return CultureInfo.CurrentUICulture.NumberFormat.NumberDecimalSeparator;
                else
                    return _decimalSeparator;
            }
            set
            {
                if (value.Length != 1)
                    throw new ArgumentException("El separador decimal debe ser un único caracter");
                _decimalSeparator = value;
            }
        }

        [Category("Appearance")]
        public int DecimalDigits {
            get
            {
                if (_decimalDigits < 0)
                    return CultureInfo.CurrentUICulture.NumberFormat.NumberDecimalDigits;
                else
                    return _decimalDigits;
            }
            set { _decimalDigits = value; }
        }

        internal NumberFormatInfo NumberFormat
        {
            get
            {
                if (_numberFormat == null)
                    _numberFormat = new NumberFormatInfo();
                _numberFormat.NumberDecimalSeparator = DecimalSeparator;
                _numberFormat.NumberDecimalDigits = DecimalDigits;
                return _numberFormat;
            }
        }

        public override object Clone()
        {
            NumericGridColumn newColumn = (NumericGridColumn)base.Clone();
            newColumn.DecimalSeparator = DecimalSeparator;
            newColumn.DecimalDigits = DecimalDigits;
            return newColumn;
        }

    }

    public class NumericGridCell: DataGridViewTextBoxCell
    {
        public NumericGridCell() : base() { }

        private NumberFormatInfo NumberFormat
        {
            get
            {
                return ((NumericGridColumn)OwningColumn).NumberFormat;
            }
        }

        public override Type ValueType
        {
            get { return typeof(decimal?); }
        }

        public override object DefaultNewRowValue
        {
            get
            {
                object defaultValue = base.DefaultNewRowValue;
                if (defaultValue is decimal)
                    return defaultValue;
                else
                    return null;
            }
        }

        public override void InitializeEditingControl(int rowIndex, object initialFormattedValue, DataGridViewCellStyle dataGridViewCellStyle)
        {
            base.InitializeEditingControl(rowIndex, initialFormattedValue, dataGridViewCellStyle);
            Control ctl = DataGridView.EditingControl;
            ctl.KeyPress += new KeyPressEventHandler(NumericCell_KeyPress);
        }

        private void NumericCell_KeyPress(object sender, KeyPressEventArgs e)
        {
            DataGridViewCell currentCell = ((IDataGridViewEditingControl)sender).EditingControlDataGridView.CurrentCell;
            if (!(currentCell is NumericGridCell)) return;

            DataGridViewTextBoxEditingControl ctl = (DataGridViewTextBoxEditingControl)sender;
            if (char.IsDigit(e.KeyChar))
            {
                int separatorPosition = ctl.Text.IndexOf(NumberFormat.NumberDecimalSeparator);
                if (separatorPosition >= 0 && ctl.SelectionStart>separatorPosition)
                {
                    int currentDecimals = ctl.Text.Length - separatorPosition 
                        - NumberFormat.NumberDecimalSeparator.Length - ctl.SelectionLength;
                    if (currentDecimals >= NumberFormat.NumberDecimalDigits)
                        e.Handled = true;
                }
            }
            else if (e.KeyChar.ToString().Equals(NumberFormat.NumberDecimalSeparator))
            {
                e.Handled = ctl.Text.IndexOf(NumberFormat.NumberDecimalSeparator) >= 0;                
            }
            else
            {
                e.Handled = !char.IsControl(e.KeyChar);
            }
        }

        public override DataGridViewCellStyle GetInheritedStyle(DataGridViewCellStyle inheritedCellStyle, int rowIndex, bool includeColors)
        {
            DataGridViewCellStyle style= base.GetInheritedStyle(inheritedCellStyle, rowIndex, includeColors);
            style.FormatProvider = NumberFormat;
            return style;
        }

        protected override object GetFormattedValue(object value, int rowIndex, ref DataGridViewCellStyle cellStyle, TypeConverter valueTypeConverter, TypeConverter formattedValueTypeConverter, DataGridViewDataErrorContexts context)
        {
            if (value == null)
                return null;
            else
                return ((decimal)value).ToString(cellStyle.FormatProvider);
        }

    }

}

File: Columna para números decimales en DataGridView/[VB]-Columna para números decimales en DataGridView/VB/GridExtension/NumericGridColumn.vb
Imports System.ComponentModel
Imports System.Globalization
Imports System.Windows.Forms

Public Class NumericGridColumn
    Inherits DataGridViewColumn

    Private _numberFormat As NumberFormatInfo
    Private _decimalSeparator As String
    Private _decimalDigits As Integer

    Public Sub New()
        MyBase.New(New NumericGridCell())
    End Sub

    Public Overrides Property CellTemplate As DataGridViewCell
        Get
            Return MyBase.CellTemplate
        End Get
        Set(value As DataGridViewCell)
            If value IsNot Nothing AndAlso
                Not value.GetType().IsAssignableFrom(GetType(NumericGridCell)) Then
                Throw New InvalidCastException("Debe especificar una instancia de NumericGridCell")
            End If
            MyBase.CellTemplate = value
        End Set
    End Property

    <Category("Appearance")>
    Public Property DecimalSeparator As String
        Get
            If String.IsNullOrEmpty(_decimalSeparator) Then
                Return CultureInfo.CurrentUICulture.NumberFormat.NumberDecimalSeparator
            Else
                Return _decimalSeparator
            End If
        End Get
        Set(ByVal value As String)
            If value.Length <> 1 Then
                Throw New ArgumentException("El separador decimal debe ser un único caracter")
            End If
            _decimalSeparator = value
        End Set
    End Property

    <Category("Appearance")>
    Public Property DecimalDigits As Integer
        Get
            If _decimalDigits < 0 Then
                Return CultureInfo.CurrentUICulture.NumberFormat.NumberDecimalDigits
            Else
                Return _decimalDigits
            End If
        End Get
        Set(ByVal value As Integer)
            _decimalDigits = value
        End Set
    End Property

    Friend ReadOnly Property NumberFormat As NumberFormatInfo
        Get
            If _numberFormat Is Nothing Then _numberFormat = New NumberFormatInfo()
            _numberFormat.NumberDecimalSeparator = DecimalSeparator
            _numberFormat.NumberDecimalDigits = DecimalDigits
            Return _numberFormat
        End Get
    End Property

    Public Overrides Function Clone() As Object
        Dim newColumn As NumericGridColumn = MyBase.Clone()
        newColumn.DecimalSeparator = DecimalSeparator
        newColumn.DecimalDigits = DecimalDigits
        Return newColumn
    End Function

End Class

Public Class NumericGridCell
    Inherits DataGridViewTextBoxCell

    Public Sub New()
        MyBase.New()
    End Sub

    Private ReadOnly Property NumberFormat As NumberFormatInfo
        Get
            Return CType(OwningColumn, NumericGridColumn).NumberFormat
        End Get
    End Property

    Public Overrides ReadOnly Property ValueType As Type
        Get
            Return GetType(Decimal?)
        End Get
    End Property

    Public Overrides ReadOnly Property DefaultNewRowValue As Object
        Get
            Dim defaultValue As Object = MyBase.DefaultNewRowValue
            If TypeOf defaultValue Is Decimal Then
                Return defaultValue
            Else
                Return Nothing
            End If
        End Get
    End Property

    Public Overrides Sub InitializeEditingControl(rowIndex As Integer, initialFormattedValue As Object, dataGridViewCellStyle As DataGridViewCellStyle)
        MyBase.InitializeEditingControl(rowIndex, initialFormattedValue, dataGridViewCellStyle)
        Dim ctl As Control = DataGridView.EditingControl
        AddHandler ctl.KeyPress, AddressOf NumericCell_KeyPress
    End Sub

    Private Sub NumericCell_KeyPress(sender As Object, e As KeyPressEventArgs)
        Dim currentCell As DataGridViewCell =
            CType(sender, IDataGridViewEditingControl).EditingControlDataGridView.CurrentCell
        If Not TypeOf currentCell Is NumericGridCell Then Return

        Dim ctl As DataGridViewTextBoxEditingControl = CType(sender, DataGridViewTextBoxEditingControl)
        If Char.IsDigit(e.KeyChar) Then
            Dim separatorPosition As Integer = ctl.Text.IndexOf(NumberFormat.NumberDecimalSeparator)
            If separatorPosition >= 0 AndAlso ctl.SelectionStart > separatorPosition Then
                Dim currentDecimals = ctl.Text.Length - separatorPosition _
                    - NumberFormat.NumberDecimalSeparator.Length - ctl.SelectionLength
                If currentDecimals >= NumberFormat.NumberDecimalDigits Then
                    e.Handled = True
                End If
            End If
        ElseIf e.KeyChar.ToString().Equals(NumberFormat.NumberDecimalSeparator) Then
            e.Handled = ctl.Text.IndexOf(NumberFormat.NumberDecimalSeparator) >= 0
        Else
            e.Handled = Not Char.IsControl(e.KeyChar)
        End If
    End Sub

    Public Overrides Function GetInheritedStyle(inheritedCellStyle As DataGridViewCellStyle, rowIndex As Integer, includeColors As Boolean) As DataGridViewCellStyle
        Dim style As DataGridViewCellStyle = MyBase.GetInheritedStyle(inheritedCellStyle, rowIndex, includeColors)
        style.FormatProvider = NumberFormat
        Return style
    End Function

    Protected Overrides Function GetFormattedValue(value As Object, rowIndex As Integer, ByRef cellStyle As DataGridViewCellStyle, valueTypeConverter As TypeConverter, formattedValueTypeConverter As TypeConverter, context As DataGridViewDataErrorContexts) As Object
        If value Is Nothing Then
            Return Nothing
        Else
            Return CType(value, Decimal).ToString(cellStyle.FormatProvider)
        End If
    End Function

End Class
